### ISSUE TITLE
Wrong detection of RedundantSelfAssignment

### ISSUE THREAD

**Issue by mrkamel** [rubocop/rubocop#12827]
Rubocop wrongly detects `Style/RedundantSelfAssignment` for e.g. a custom `append` method which is not self assigning, introducing bugs when rubocop autocorrect is used.

--------

## Expected behavior

Rubocop should not detect and fix `Style/RedundantSelfAssignment` for custom `append`/`concat`/etc methods. 
In my opinion `append`, `concat`, etc. are way too generic method names and do not even contain a bang in the name and therefore are often used/defined in a non-self assigning manner. I stumbled upon it when i wanted to add a `append` method to my gem https://github.com/mrkamel/kraps as a non-self-assigning method. Even if i'd ignore this in my gem, users of the gem would also easily be affected by this issue. Futher, rubocop detection based only on a method name is a very error-prone approach in general, imho.

## Actual behavior

Rubocop wrongly detects it and rubocop autocorrect removes the assignment, resulting in a bug.

## Steps to reproduce the problem

Given this example code:

```ruby
# frozen_string_literal: true

# MyJob
class MyJob
  def append(other)
    [self, other]
  end 
end

job = MyJob.new
job = job.append(MyJob.new)

puts job 
```
and runing rubocop

```
For ... Default configuration from ...rbenv/versions/3.1.0/lib/ruby/gems/3.1.0/gems/rubocop-1.63.0/config/default.yml
Use parallel by default.
Skipping parallel inspection: only a single file needs inspection
Inspecting 1 file
Scanning .../test.rb
Loading cache from ...
C

Offenses:

test.rb:11:3: C: [Correctable] Style/RedundantSelfAssignment: Redundant self assignment detected. Method append modifies its receiver in place.
job = job.append(MyJob.new)
  ^

1 file inspected, 1 offense detected, 1 offense autocorrectable
Finished in 0.11157057499985967 seconds
```

## RuboCop version

```
$ rubocop -V
1.63.0 (using Parser 3.3.0.5, rubocop-ast 1.31.2, running on ruby 3.1.0) [x86_64-linux]
```

**Comment by koic** [comment]
This cop is already marked as unsafe:
https://docs.rubocop.org/rubocop/1.62/cops_style.html#styleredundantselfassignment

 If there are many reports of this nature, considering disabling it by default might be an option, but in actuality, there hasn’t been any notable issue feedback to date. Observing future feedbacks to make a decision would be prudent. Thank you.
